### ISSUE TITLE
Agents Page should not allow users to modify Environments associated through Config Repo

### ISSUE THREAD

**Issue by akshaydewan** [gocd/gocd#5861]
Related to #5761

### Summary
If an Agent is associated with an Environment via a Config Repo, it should not be editable (meaning, a user shouldn't be able to update the agent and un-select the Envrionment). The current UI makes this possible, but results in a no-op with a success message.

### Implementation
The Agents API returns a model that contains only the associated environment names as strings. We will need to add a flag saying whether the environment was associated from a Config Repo.

Current response example:
```json
{
  "_links" : { //..snipped
  },
  "_embedded" : {
    "agents" : [ {
      "_links" : { //..snipped
      },
      "uuid" : "1c7982c8-6615-492f-871b-1a98f70c08e5",
      "hostname" : "in-akshaydewan.local",
      "ip_address" : "127.0.0.1",
      "sandbox" : "",
      "operating_system" : "",
      "agent_config_state" : "Enabled",
      "agent_state" : "LostContact",
      "environments" : [ "dev" ],
      "build_state" : "Unknown",
      "free_space" : "unknown",
      "resources" : [ ]
    }, {
      "_links" : { //..snipped
      },
      "uuid" : "f7ba19f9-21f6-4120-aec1-db6d4564a115",
      "hostname" : "in-akshaydewan.local",
      "ip_address" : "127.0.0.1",
      "sandbox" : "",
      "operating_system" : "",
      "agent_config_state" : "Enabled",
      "agent_state" : "LostContact",
      "environments" : [ ],
      "build_state" : "Unknown",
      "free_space" : "unknown",
      "resources" : [ ]
    } ]
  }
}
```

The change in the environments object can be:
```json
{
  "environments": [
    {
      "name": "dev",
      "origin": {
        "type": "config-repo",
        "_links": {
          "self": {
            "href": "http://localhost:8153/go/api/admin/config_repos/foo"
          },
          "doc": {
            "href": "https://api.gocd.org/#config-repos"
          },
          "find": {
            "href": "http://localhost:8153/go/api/admin/config_repos/:id"
          }
        }
      }
    }
  ]
}
```

It will require the Agents API version to be bumped up.

There is another call which is made to an internal Environments endpoint `/go/api/admin/internal/environments`, but we don't send any Agents' information here, and IMO, it makes more sense to add the information to the Agent.

### Question
Let's say we have two agents and two environments. Agent1 is associated with Env1 from a Config Repo, but Agent2 is not. Now, selecting multiple agents in the UI will result in an indeterminate Schrödinger's checkbox, which should be enabled and disabled at the same time. What should be done in this case? I think we can make it inderminate and disabled.
<img width="1440" alt="screenshot 2019-02-15 at 16 43 18" src="https://user-images.githubusercontent.com/6024038/52853184-d3a53b80-3140-11e9-8535-4e2f306cf397.png">

**Comment by akshaydewan** [comment]
cc: @maheshp @arvindsv

**Comment by arvindsv** [comment]
> Schrödinger's checkbox

🤣 

Yes, indeterminate and disabled sounds alright to me.

**Comment by akshaydewan** [comment]
We can end up with an invalid configuration when selecting/unselecting environments on the Agents page.

### Steps to reproduce
1. Create two environments - one defined in XML (envXml) and one in a ConfigRepo (envCR).
2. Create two agents, A1 and A2. A1 is associated with envCR in the Config Repo.
3. Select all agents on the Agents page and associate envXml with both agents.
4. Select all agents and un-select envXml for both agents.
Result: The association of A1 with envCR will end up in the ConfigXML, even though it is already associated in the ConfigRepo. When this invalid state is reached, the agents API will break, or the Config Repo will fail to parse after an MDU, saying that the agent is defined more than once.
Expected result: The association of A1 with envCR should not be written to the configXML.

The payload being sent to the server on Step 4 is:
```json
{
  "uuids": [
    "07717f78-cb17-4a1f-b837-cf01a88710ff",
    "f78c4d0a-93a2-4170-b537-17d818387943"
  ],
  "operations": {
    "environments": {
      "add": [
        "envCR"
      ],
      "remove": [
        "envXML"
      ]
    }
  }
}
```

**This happens even when the checkbox is disabled, as implemented in PR #5886**

### Possible fix
We should send the right payload from the UI, and also validate the operation from on the backend to make sure we don't end up with an invalid config.

**Comment by rajiesh** [comment]
Verified on 19.3.0 (8959-259d1f79e361875995072e5c1fcc9cbda7ccc600)
